azurerm/resource_arm_virtual_machine_migration.go

```diff
@@ -0,0 +1,85 @@
+package azurerm
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func resourceAzureRMVirtualMachineMigrateState(
+	v int, is *terraform.InstanceState, meta interface{}) (*terraform.InstanceState, error) {
+	switch v {
+	case 0:
+		log.Println("[INFO] Found AzureRM Virtual Machine State v0; migrating to v1")
+		return migrateAzureRMVirtualMachineStateV0toV1(is)
+	default:
+		return is, fmt.Errorf("Unexpected schema version: %d", v)
+	}
+}
+
+func migrateAzureRMVirtualMachineStateV0toV1(is *terraform.InstanceState) (*terraform.InstanceState, error) {
+	if is.Empty() {
+		log.Println("[DEBUG] Empty InstanceState; nothing to migrate.")
+		return is, nil
+	}
+
+	log.Printf("[DEBUG] ARM Virtual Machine Attributes before Migration: %#v", is.Attributes)
+
+	if is.Attributes["os_profile_windows_config.#"] == "1" {
+		hash, err := computeAzureRMVirtualMachineStateV1WinConfigHash(is, "UTC")
+		if err != nil {
+			return is, fmt.Errorf("Failed to calculate the new hash code for os_profile_windows_config field: %#v", err)
+		}
+		replaceAzureRMVirtualMachineStateV0WinConfigWithNewHash(is, hash)
+		tzKey := fmt.Sprintf("%s.%d.%s", "os_profile_windows_config", hash, "timezone")
+		is.Attributes[tzKey] = "UTC"
+	}
+
+	log.Printf("[DEBUG] ARM Virtual Machine Attributes after State Migration: %#v", is.Attributes)
+
+	return is, nil
+}
+
+func computeAzureRMVirtualMachineStateV1WinConfigHash(is *terraform.InstanceState, timezone string) (int, error) {
+	data := make(map[string]interface{})
+	for k, v := range is.Attributes {
+		if !strings.HasPrefix(k, "os_profile_windows_config.") {
+			continue
+		}
+		paths := strings.Split(k, ".")
+		if len(paths) == 3 && (paths[2] == "enable_automatic_upgrades" || paths[2] == "provision_vm_agent") {
+			b, err := strconv.ParseBool(v)
+			if err != nil {
+				return 0, err
+			}
+			data[paths[2]] = b
+		}
+	}
+	data["timezone"] = timezone
+	return resourceArmVirtualMachineStorageOsProfileWindowsConfigHash(data), nil
```

Comment:
The underlying data structures will diverge significantly in a future version of Terraform - as such direct manipulation of the state will cause hard to diagnose crashes later. Instead we need to use the schema parser ([example](https://github.com/terraform-providers/terraform-provider-azurerm/blob/master/azurerm/resource_arm_container_registry_migrate.go#L61)) rather than assuming the format for keys, (as here)

This means we can simply assign the new value for the field, and allow Terraform Core to recompute the hash value; meaning the tests _should_ stay the same, but all of this can disappear

Reply:
(heads up that this needs to be resolved before we can merge this, which is why this is still "requested changes" 😄)

Just to add some additional context to testing this; you'll want to compare the state of a VM using the current provider release (currently v1.6.0) to the current development build you're working on; for instance:


```hcl
provider "azurerm" {
  version = "=1.6.0"
}

resource "azurerm_resource_group" "test" {
  # ..
}

# ...
```

```bash
$ rm -rf .terraform/
$ terraform init
$ terraform apply
```

then un-pinning the provider version and re-initing/applying should show the state migration:

```hcl
provider "azurerm" {
}

resource "azurerm_resource_group" "test" {
  # ..
}

# ...
```

```bash
$ rm -rf .terraform/
$ terraform init # to use the development version you've built via `make`
$ terraform plan|apply
```

In order to test this effectively unfortunately you'll want to destroy & recreate the VM for each iteration, which the `taint` command is quite helpful for:

```
$ terraform taint azurerm_virtual_machine.test
$ # switch back to "1.6.0"
$ terraform apply
```

Hope that helps :)

Reply:
The test steps is really helpful here. So AFAIK, we can only do the manual test here for this kind of issues? Are there any unit-test infrastructure support it?

Reply:
> So AFAIK, we can only do the manual test here for this kind of issues? Are there any unit-test infrastructure support it?

The associated unit test is checking the logic from a given version - this is more exploratory testing